[10.0][ADD] product service duration

woodbrettm opened the pull request:
Hi there,

Opened in response to #206.

### Description

Extension of products (services only) that allows them to be added to calendars.

As a note, if you add services with a `Minimum Time` (minimum time it
takes to complete that service) to an event, the meeting/event duration cannot
be less than the combined minimum time of those services.

For example, say your meeting involves two services, technical and design services.
If the technical service requires 2 hours and design requires 3 hours, then your
meeting must be at least 5 hours.

The above constraint does not apply on events that have `All Day` selected.

When changing the `Minimum Time` on services, this will not affect events that have
already ended.



Thanks!
Brett

Commit:
[ADD] Add module product_service_duration.

woodbrettm commented on product_service_duration/demo/calendar_event_demo.xml:
```diff
@@ -0,0 +1,17 @@
+        <field name="duration" eval="5.00" />
+        <field name="template_ids"
+               eval="[(4, ref('product_service_duration.product_template_1')),
+                      (4, ref('product_service_duration.product_template_2'))]" />
```
Pretty sure I can do [(6, 0 [ids])] right?

lasley replied:
Yup, that's actually what you want to do. Subsequent updates of this record will actually continue to insert the same records into the m2m, because the old ones aren't being removed. `6` is just a `5` followed by a `4` for each `id` in `ids`

woodbrettm commented on product_service_duration/models/calendar_event.py:
```diff
@@ -0,0 +1,45 @@
+                event_hours = diff_secs / 3600
+
+            min_event_duration = sum(
+                [t.minimum_service_time for t in record.template_ids]
```
Perhaps I should add `min_event_duration` as a computed field. Would be more useful for other devs that way.

lasley replied:
That's fine, yeah. Just make sure to store it.

woodbrettm commented on product_service_duration/views/product_template_view.xml:
```diff
@@ -0,0 +1,44 @@
+            <xpath expr="//group[@name='group_standard_price']" position="after">
+                <group name="group_service">
+                    <field name="service_user_id" />
+                    <field name="minimum_service_time" />
```
Should probably default these to invisible if the product is not a service.

woodbrettm commented on product_service_duration/models/calendar_event.py:
```diff
@@ -0,0 +1,45 @@
+                )
+
+                diff_secs = (stop - start).total_seconds()
+                event_hours = diff_secs / 3600
```
Should I add any rounding to this? Not sure if that may cause issues when comparing event_hours against the duration of the meeting, since they could potentially have a different set of decimal points.

lasley replied:
You should never round without an explicit reason, and definitely not before saving data. Rounding should always be a display thing

woodbrettm commented:
Travis is failing from unused imports at the moment. Will fix after reviews.

lasley commented on product_service_duration/models/calendar_event.py:
```diff
@@ -0,0 +1,45 @@
+            )
+            if min_event_duration > event_hours:
+                raise ValidationError(_(
+                    'The services involved requires a %s hour '
```
s/requires/require

lasley commented on product_service_duration/models/calendar_event.py:
```diff
@@ -0,0 +1,45 @@
+            if min_event_duration > event_hours:
+                raise ValidationError(_(
+                    'The services involved requires a %s hour '
+                    'long meeting.' % min_event_duration,
```
This doesn't work. You're interpolating on the second string, not the first. 

Regardless, interpolation should take place outside of the translation. Otherwise you need a new translation for any variable combination.
```
_('this is data %s') % data
```

lasley commented on product_service_duration/models/product_template.py:
```diff
@@ -0,0 +1,28 @@
+
+    service_user_id = fields.Many2one(
+        string='Service User',
+        comodel_name='res.users',
```
`domain="[('company_id', 'child_of', company_id)]"`

woodbrettm replied:
What's the reason for this domain?

lasley replied:
This is not multi company friendly as it is. A company would only be selling services for their users.

woodbrettm replied:
oh got it. Makes sense now.

lasley commented on product_service_duration/models/product_template.py:
```diff
@@ -0,0 +1,28 @@
+    )
+    minimum_service_time = fields.Float(
+        string='Minimum Service Hours',
+        digits_compute=dp.get_precision('Product UoM'),
```
[s/digits_compute/digits](https://github.com/odoo/odoo/search?utf8=%E2%9C%93&q=%22dp.get_precision%28%22&type=)

lasley commented on product_service_duration/tests/test_calendar_event.py:
```diff
@@ -0,0 +1,86 @@
+import time
+
+
+class TesteventEvent(TransactionCase):
```
Bad class name

lasley reviewed (changes requested):
Thanks @bmw95

Commit:
[IMP] product_service_duration: Refactor models and views.
* Change any time fields to float_time widgets
* Adjust field names
* Add domain to service_user_id
* Abstract calendar_event logic into smaller chunks

woodbrettm commented:
@lasley updated.

lasley commented on product_service_duration/views/calendar_event_view.xml:
```diff
@@ -0,0 +1,58 @@
+            <xpath expr="//field[@name='partner_ids']" position="after">
+                <field name="template_ids"
+                       widget="many2many_tags"
+                       can_create="true"
```
domain restrict to only service type products

woodbrettm replied:
Hmm why type products? I think earlier in the review you specified in another template ids field in the calendar popup form to do product type service. Which one should I do?

lasley replied:
products with `type == 'service'` is what I meant in both areas.

lasley commented on product_service_duration/models/product_template.py:
```diff
@@ -0,0 +1,29 @@
+
+    _inherit = 'product.template'
+
+    service_user_id = fields.Many2one(
```
This actually needs to be to a resource, not to a user. The resource is the user in the context of their working hours.

woodbrettm replied:
That makes a lot more sense now haha

lasley commented on product_service_duration/models/product_template.py:
```diff
@@ -0,0 +1,29 @@
+import odoo.addons.decimal_precision as dp
+
+
+class ProductTemplate(models.Model):
```
This actually needs to be pushed back into the variant please. Think multiple doctors that perform the same exam types.

Commit:
[IMP] product_service_duration: Add new logic checks.
* Prevent validation errors if event allday
* Prevent validation errors if event stop date is in the past
* Add test coverage
* Move product logic to product.product
* Move view to product product views

woodbrettm commented:
@lasley thanks for the review

I updated some of the logic to prevent validation errors if an event has ended as well as if the event is selected as all_day.

With the above bit in mind, I also added some logic to prevent users from changing the minimum service time on products if they're causing conflicts in events.

A dependency on `sale` was added as well, as the product.product views only seem to be located in the product variants side menu.

Cheers!

lasley commented:
> With the above bit in mind, I also added some logic to prevent users from changing the minimum service time on products if they're causing conflicts in events.

Previously scheduled events should not be changing if the service time on the product is changing. It should only be relevant for new appointments

Commit:
[FIX] product_service_duration: Fix views, event model.
* Stop min_duration from being calculated if event all day or event ended
* Add event min_duration to calendar form popup
* Add event min_duration to calendar search

woodbrettm commented:
@lasley I think I may actually be misunderstanding, however to confirm, I'm now using the following logic for both the validations and compute_min_duration in calendar event:
```python
            if record.stop < today_time:
                continue

            if record.allday:
                continue
```
So the logic will only apply to currently running events and events in the future. 
Is that good?

lasley reviewed (approved):
Thanks @bmw95

lasley commented:
@tedsalmon please review

tedsalmon reviewed (approved):
LGTM 👍
